core/src/main/scala/org/apache/spark/CacheManager.scala

```diff
@@ -47,9 +49,13 @@ private[spark] class CacheManager(blockManager: BlockManager) extends Logging {
         val inputMetrics = blockResult.inputMetrics
         val existingMetrics = context.taskMetrics
           .getInputMetricsForReadMethod(inputMetrics.readMethod)
-        existingMetrics.addBytesRead(inputMetrics.bytesRead)
+        existingMetrics.incBytesRead(inputMetrics.bytesRead)
 
-        new InterruptibleIterator(context, blockResult.data.asInstanceOf[Iterator[T]])
+        val iter = blockResult.data.asInstanceOf[Iterator[T]]
+        new InterruptibleIterator(context, AfterNextInterceptingIterator(iter, (next: T) => {
+          existingMetrics.incRecordsRead(1)
```

Comment:
The question is - how expensive is the thing we are doing inside of the override method? In those other cases I think we're just checking the value of a single variable that doesn't change often (i.e. checking for interrupted). In the past we've seen performance regressions from anything more expensive than this: See https://github.com/apache/spark/commit/f708dda81ed5004325591fcc31cd79a8afa580db.

The cost of CAS is hardware dependent, but can be expensive on machines with large numbers of cores because in many case there is a single shared bus lock. Volatile is similar, basically if you think about it maybe you have 16 cores and they each need to constantly invalidate each-other's local copy of the variable.